File: engine/__tests__/monitoring.test.ts
import { generateMonitoringPlan } from '../monitoring';

describe('monitoring schedule generator', () => {
  test('lithium baseline includes U&E + TFT + Ca + ECG', () => {
    const plan = generateMonitoringPlan({ toDrugId: 'lithium' });
    const labels = plan.entries.map((e) => e.label);
    expect(labels).toEqual(expect.arrayContaining(['U&E + eGFR', 'TFT', 'Calcium', 'ECG']));
  });

  test('lithium recurring level expands at 90-day interval', () => {
    const plan = generateMonitoringPlan({ toDrugId: 'lithium', durationDays: 365 });
    const days = plan.entries
      .filter((e) => e.label === 'Lithium level')
      .map((e) => e.dayOffset);
    // Should include the initial 7-day, then 90/180/270/360
    expect(days).toEqual(expect.arrayContaining([7, 90, 180, 270]));
  });

  test('clozapine generates weekly FBC for 18 weeks', () => {
    const plan = generateMonitoringPlan({ toDrugId: 'clozapine', durationDays: 365 });
    const weeklyFbc = plan.entries.filter((e) => e.label === 'Weekly FBC');
    // 18 weeks × weekly = 18 entries
    expect(weeklyFbc.length).toBe(18);
  });

  test('antidepressant fallback includes mood/suicidality review at D14', () => {
    const plan = generateMonitoringPlan({ toDrugId: 'sertraline' });
    const review = plan.entries.find(
      (e) => e.label === 'Mood + suicidality' && e.dayOffset === 14,
    );
    expect(review).toBeDefined();
  });

  test('cardiac comorbidity adds baseline ECG even without QT-prolonger', () => {
    const plan = generateMonitoringPlan({
      toDrugId: 'mirtazapine',
      context: { ageYears: 50, sex: 'male', comorbidities: { cardiac: true } },
    });
    expect(plan.entries.some((e) => e.label === 'ECG (cardiac hx)')).toBe(true);
  });

  test('plan has citations and a non-zero span', () => {
    const plan = generateMonitoringPlan({ toDrugId: 'lithium' });
    expect(plan.citations.length).toBeGreaterThan(0);
    expect(plan.spanDays).toBeGreaterThan(0);
  });

  test('entries are sorted by dayOffset', () => {
    const plan = generateMonitoringPlan({ toDrugId: 'olanzapine' });
    for (let i = 1; i < plan.entries.length; i++) {
      expect(plan.entries[i].dayOffset).toBeGreaterThanOrEqual(plan.entries[i - 1].dayOffset);
    }
  });
});

File: flutter_app/test/engine/monitoring_test.dart
// Tests for the Dart monitoring port.
// Mirrors engine/__tests__/monitoring.test.ts.

import 'package:flutter_test/flutter_test.dart';
import 'package:psychswitch/src/engine/monitoring.dart';
import 'package:psychswitch/src/engine/patient_context_pure.dart';

void main() {
  group('monitoring schedule generator', () {
    test('lithium baseline includes U&E + TFT + Ca + ECG', () {
      final plan = generateMonitoringPlan(toDrugId: 'lithium');
      final labels = plan.entries.map((e) => e.label).toList();
      expect(labels, containsAll(<String>['U&E + eGFR', 'TFT', 'Calcium', 'ECG']));
    });

    test('lithium recurring level expands at 90-day interval', () {
      final plan = generateMonitoringPlan(
        toDrugId: 'lithium',
        durationDays: 365,
      );
      final days = plan.entries
          .where((e) => e.label == 'Lithium level')
          .map((e) => e.dayOffset)
          .toList();
      expect(days, containsAll(<int>[7, 90, 180, 270]));
    });

    test('clozapine generates weekly FBC for 18 weeks', () {
      final plan = generateMonitoringPlan(
        toDrugId: 'clozapine',
        durationDays: 365,
      );
      final weeklyFbc =
          plan.entries.where((e) => e.label == 'Weekly FBC').toList();
      expect(weeklyFbc.length, equals(18));
    });

    test('antidepressant fallback includes mood/suicidality review at D14', () {
      final plan = generateMonitoringPlan(toDrugId: 'sertraline');
      expect(
        plan.entries.any(
          (e) => e.label == 'Mood + suicidality' && e.dayOffset == 14,
        ),
        isTrue,
      );
    });

    test('cardiac comorbidity adds baseline ECG even without QT-prolonger', () {
      final plan = generateMonitoringPlan(
        toDrugId: 'mirtazapine',
        context: const PatientContext(
          ageYears: 50,
          sex: Sex.male,
          comorbidities: Comorbidities(cardiac: true),
        ),
      );
      expect(
        plan.entries.any((e) => e.label == 'ECG (cardiac hx)'),
        isTrue,
      );
    });

    test('plan has citations and a non-zero span', () {
      final plan = generateMonitoringPlan(toDrugId: 'lithium');
      expect(plan.citations, isNotEmpty);
      expect(plan.spanDays, greaterThan(0));
    });

    test('entries are sorted by dayOffset', () {
      final plan = generateMonitoringPlan(toDrugId: 'olanzapine');
      for (var i = 1; i < plan.entries.length; i++) {
        expect(
          plan.entries[i].dayOffset,
          greaterThanOrEqualTo(plan.entries[i - 1].dayOffset),
        );
      }
    });

    test('unknown drug returns empty plan', () {
      final plan = generateMonitoringPlan(toDrugId: 'not-a-drug');
      expect(plan.entries, isEmpty);
    });
  });

  group('MonitoringCategory jsonValue round-trips', () {
    test('every category parses back', () {
      for (final c in MonitoringCategory.values) {
        expect(MonitoringCategory.fromJson(c.jsonValue), equals(c));
      }
    });
  });

  group('MonitoringEntry.toJson', () {
    test('omits drugId/citation/recurring when null', () {
      const e = MonitoringEntry(
        dayOffset: 0,
        label: 'X',
        detail: 'd',
        category: MonitoringCategory.lab,
      );
      final j = e.toJson();
      expect(j.containsKey('drugId'), isFalse);
      expect(j.containsKey('citation'), isFalse);
      expect(j.containsKey('recurring'), isFalse);
    });

    test('includes recurring with everyDays + optional untilDay', () {
      const e = MonitoringEntry(
        dayOffset: 7,
        label: 'Weekly X',
        detail: 'd',
        category: MonitoringCategory.lab,
        recurring: MonitoringRecurrence(everyDays: 7, untilDay: 28),
      );
      final j = e.toJson();
      expect(
        j['recurring'],
        equals(<String, dynamic>{'everyDays': 7, 'untilDay': 28}),
      );
    });
  });
}
